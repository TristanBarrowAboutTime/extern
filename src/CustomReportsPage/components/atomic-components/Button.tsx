import * as React from 'react';
import { FunctionComponent } from 'react';
import styled from 'styled-components';
import { ButtonType } from '../../types/ButtonType';
import { useButton } from '../../hooks/component-hooks/atomic-components/useButton';

type StyledButtonProps = {
    backgroundColor: string
    shadow: string
    padding: string
    cursor: string
    color: string
}

const StyledButton = styled.div`
    border: 0;
    border-radius: 4px;
    white-space: nowrap;
    background-color: ${(props: StyledButtonProps) => props.backgroundColor};
    color: ${(props: StyledButtonProps) => props.color};
    box-shadow: ${(props: StyledButtonProps) => props.shadow};
    padding: ${(props: StyledButtonProps) => props.padding};
    :hover {
        cursor: ${(props: StyledButtonProps) => props.cursor}
    }
    :focus {
        border: 0;
    }
`;

type ButtonProps = {
    buttonType?: ButtonType,
    text: string,
    styles?: object,
    onClick: () => void,
    disabled?: boolean,
}

const Button: FunctionComponent<ButtonProps> = ({
    buttonType = ButtonType.NORMAL, 
    text,
    onClick,
    styles = {},
    disabled = false
}: ButtonProps) => {
    const binding = useButton({buttonType, disabled, onClick});

    return (
        <StyledButton
            style={styles}
            onClick={binding.click}
            backgroundColor={binding.style.backgroundColor}
            padding={binding.style.padding}
            color={binding.style.color}
            shadow={binding.style.shadow}
            cursor={binding.style.cursor}
        >
            {text}
        </StyledButton>
    );
}

export default Button;