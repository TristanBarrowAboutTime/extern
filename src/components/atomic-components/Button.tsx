import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import Styles from '../../style/Styles';

/**
 * This will probably need to be turned into react-native at some point.
 */
export enum ButtonType {
    SMALL_NORMAL = 'SMALL_NORMAL',
    SMALL_GREEN = 'SMALL_GREEN',
    SMALL_BARE = 'SMALL_BARE',
    SMALL_RED = 'SMALL_RED',
    NORMAL = 'NORMAL',
    GREEN = 'GREEN',
    BARE = 'BARE',
    RED = 'RED',
}

type StyledButtonProps = {
    buttonType: ButtonType
    disabled: boolean
}

const getBgColor = (props: StyledButtonProps): string => {
    if (props.disabled) {
        switch(props.buttonType) {
            case ButtonType.NORMAL: 
            case ButtonType.SMALL_NORMAL:
            case ButtonType.GREEN:  
            case ButtonType.SMALL_GREEN:
            case ButtonType.RED:
            case ButtonType.SMALL_RED:
                return Styles.color.gray.medium;
            case ButtonType.BARE:
            case ButtonType.SMALL_BARE:
                return Styles.color.transparent;
            default:
                return Styles.color.gray.medium;
        }
    } else {
        switch(props.buttonType) {
            case ButtonType.NORMAL: 
            case ButtonType.SMALL_NORMAL:
                return Styles.color.gray.xx_dark;
            case ButtonType.GREEN:  
            case ButtonType.SMALL_GREEN:
                return Styles.color.green;
            case ButtonType.RED:
            case ButtonType.SMALL_RED:
                return Styles.color.red;
            case ButtonType.BARE:
            case ButtonType.SMALL_BARE:
                return Styles.color.transparent;
            default: 
                return Styles.color.gray.xx_dark;
        }
    }
}

const getPadding = (type: ButtonType) => {
    switch(type) {
        case ButtonType.SMALL_NORMAL: 
        case ButtonType.SMALL_GREEN:
        case ButtonType.SMALL_RED:
        case ButtonType.SMALL_BARE:
            return Styles.button.paddingSmall;
        case ButtonType.NORMAL: 
        case ButtonType.GREEN:  
        case ButtonType.RED:
        case ButtonType.BARE:
            return Styles.button.paddingNormal;
        default:
            return '10px'; 
    } 
    
}

const getTextColor = (props: StyledButtonProps): string => {
    if (props.disabled) {
        switch(props.buttonType) {
            case ButtonType.NORMAL:
            case ButtonType.SMALL_NORMAL:
            case ButtonType.GREEN:
            case ButtonType.SMALL_GREEN:
            case ButtonType.RED:
            case ButtonType.SMALL_RED:
                return Styles.color.white;
            case ButtonType.BARE:
            case ButtonType.SMALL_BARE:
                return Styles.color.gray.dark;
            default: 
                return Styles.color.white;
        }
    } else {
        switch(props.buttonType) {
            case ButtonType.NORMAL: 
            case ButtonType.SMALL_NORMAL:
            case ButtonType.GREEN: 
            case ButtonType.SMALL_GREEN:
            case ButtonType.RED:
            case ButtonType.SMALL_RED:
                return Styles.color.white;
            case ButtonType.BARE:
            case ButtonType.SMALL_BARE:
                return Styles.color.gray.xx_dark;
            default: 
                return Styles.color.gray.xx_dark;
        }
    }
}

const getShadow = (props: StyledButtonProps): string => {
    if (props.buttonType === ButtonType.BARE || 
        props.buttonType === ButtonType.SMALL_BARE || 
        props.disabled) 
    { 
        return '0'; 
    }
    else return Styles.button.shadow;

}

// eventually this implementation should be changed to something more sustainable
const Container = styled.button`
    border: 0;
    border-radius: 4px;
    background-color: ${(props: StyledButtonProps) => getBgColor(props)};
    color: ${(props: StyledButtonProps) => getTextColor(props)};
    box-shadow: ${(props: StyledButtonProps) => getShadow(props)};
    padding: ${(props: StyledButtonProps) => getPadding(props.buttonType)};
    :hover {
        cursor: ${(props: StyledButtonProps) => props.disabled ? 'default' : 'pointer'}
    }
    :focus {
        border: 0;
    }
`;

type ButtonProps = {
    buttonType?: ButtonType,
    text: string,
    styles?: object,
    onClick: () => void,
    disabled?: boolean,
}

export const Button: FunctionComponent<ButtonProps> = ({
    buttonType = ButtonType.NORMAL, 
    text,
    onClick,
    styles = {},
    disabled = false
}: ButtonProps) => {

    return (
        <Container 
            style={styles}
            onClick={onClick}
            disabled={disabled}
            buttonType={buttonType}
        >
            {text}
        </Container>
    );
}


export default Button;