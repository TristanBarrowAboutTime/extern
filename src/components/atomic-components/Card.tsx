import React, { ReactNode } from 'react';
import styled from 'styled-components';

type StyleProps = {
}

const Container = styled.div`
    display: flex;
    background-color: white;
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.25);
    box-sizing: border-box;
    border-radius: 4px;
    margin: 8px;
    padding: 16px;
    transition: all .2s ease;
    width: calc(100%/2 - 16px);

    @media only screen and (min-width: 1150px) {
        width: calc(100%/3 - 16px);    
    }

    @media only screen and (max-width: 850px) {
        width: auto;
        flex-grow: 1;
    }

    
`;

type CardProps = {
    children: ReactNode
}

const Card = (props: CardProps) => {
    return (
        <Container>
            {props.children}
        </Container>
    );
}

export default Card; 