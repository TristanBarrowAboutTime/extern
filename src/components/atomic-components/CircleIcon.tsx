import React from 'react';
import { View, StyleSheet, TouchableWithoutFeedback } from 'react-native';
import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Styles from '../../style/WMStyles'

type CircleIconProps = {
    icon: IconProp 
    color?: string
    bgColor?: string
    size?: number
    marginLeft?: number
    marginRight?: number
    circleSize?: number
    onPress?: () => void
}

const CircleIcon = ({
    icon,
    color = Styles.color.white,
    bgColor = Styles.color.gray.x_dark,
    size = 11,
    circleSize = 16,
    marginLeft = 0,
    marginRight = 0,
    onPress
}: CircleIconProps) => {
    const styles = StyleSheet.create({
        iconContainer: {
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
            height: circleSize,
            width: circleSize,
            borderRadius: 100/2,
            backgroundColor: bgColor,
            fontSize: size,
            marginLeft: marginLeft,
            marginRight: marginRight,
            marginTop: 0,
            marginBottom: 0,
            '&:hover': {
                cursor: 'pointer'
            }
        }
});
    return (
        <TouchableWithoutFeedback onPress={onPress}>
            <View style={styles.iconContainer}>
                <FontAwesomeIcon icon={icon} color={color} />
            </View>
        </TouchableWithoutFeedback>
    );
} 



export default CircleIcon;