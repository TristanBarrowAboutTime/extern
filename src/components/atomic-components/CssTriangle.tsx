import React from 'react';
import styled from 'styled-components';

type StyleProps = {
    size: number,
    color: string,
}

const defaultColor = '#000000';
const defaultSize = 5;

const UpArrow = styled.div`
    width: 0;
    height: 0;
    border-left: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-right: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-bottom: ${(props: StyleProps) => `${props.size}px solid ${props.color}`};
`;

const DownArrow = styled.div`
    width: 0;
    height: 0;
    border-left: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-right: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-top: ${(props: StyleProps) => `${props.size}px solid ${props.color}`};
`;

const LeftArrow = styled.div`
    width: 0;
    height: 0;
    border-top: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-bottom: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-right: ${(props: StyleProps) => `${props.size}px solid ${props.color}`};
`;

const RightArrow = styled.div`
    width: 0;
    height: 0;
    border-top: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-bottom: ${(props: StyleProps) => `${props.size}px solid transparent`};
    border-left: ${(props: StyleProps) => `${props.size}px solid ${props.color}`};
`;

type CssTriangleProps = {
    className?: string
    onClick?: () => void
    color?: string
    size?: number
}

export const Up = ({
    className = '',
    onClick,
    color = defaultColor,
    size = defaultSize,
}: CssTriangleProps) => {
    return (
        <div className={className} onClick={onClick}>
            <UpArrow size={size} color={color} />
        </div>
    );
}

export const Down = ({
    className = '',
    onClick,
    color = defaultColor,
    size = defaultSize,
}: CssTriangleProps) => {
    return (
        <div className={className} onClick={onClick}>
            <DownArrow size={size} color={color} />
        </div>
    );
}

export const Left = ({
    className = '',
    onClick,
    color = defaultColor,
    size = defaultSize,
}: CssTriangleProps) => {
    return (
        <div className={className} onClick={onClick}>
            <LeftArrow size={size} color={color} />
        </div>
    );
}

export const Right = ({
    className = '',
    onClick,
    color = defaultColor,
    size = defaultSize,
}: CssTriangleProps) => {
    return (
        <div className={className} onClick={onClick}>
            <RightArrow size={size} color={color} />
        </div>
    );
}