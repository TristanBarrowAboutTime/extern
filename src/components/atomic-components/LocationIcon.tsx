import * as React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faMapMarker } from '@fortawesome/free-solid-svg-icons';
import styled from 'styled-components';
import { useInitials } from '../../hooks/useInitials';

const Container = styled.div`
    position: relative;
`; 

type InitialsStyle = {
    size: number
    color: string
}

const Initials = styled.div`
    position: absolute;
    text-align: center;
    box-sizing: border-box;
    height: ${(props: InitialsStyle) => props.size}px;
    width: ${(props: InitialsStyle) => props.size * 0.75}px;
    font-size: ${(props: InitialsStyle) => props.size * 0.4}px;
    padding-top: ${(props: InitialsStyle) => props.size * 0.1}px;
    color: ${(props: InitialsStyle) => props.color};
    
`;


type LocationIconProps = {
    first: string
    last: string
    color?: string
    textColor?: string
    size?: number
}

const LocationIcon = (props: LocationIconProps) => {
    const { first, last, color, size, textColor } = props;

    const initals = useInitials({ first, last })
    return (
        <Container>
            <Initials 
                size={size as number} 
                color={textColor as string}
            >
                {initals}
            </Initials>
            <FontAwesomeIcon
                icon={faMapMarker} 
                color={color} 
                style={{fontSize: size}} 
            />
        </Container>

    ); 
}

LocationIcon.defaultProps = {
    color: 'gray',
    textColor: '#E9E9E9',
    size: 50,

}

export default LocationIcon;