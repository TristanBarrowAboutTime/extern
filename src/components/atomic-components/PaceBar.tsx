import React from 'react';
import { useMemo } from 'react'; 
import styled from 'styled-components';

type ProgressStyleProps = {
    width: string
    color: string
}

const FullBar = styled.div`
    position: relative;
    box-sizing: border-box;
    width: 100%;
    height: 24px;
    border: 1px solid black;
`;

const Progress = styled.div`
    background-color: ${(props: ProgressStyleProps) => props.color};
    width: ${(props: ProgressStyleProps) => props.width};
    height: 22px;
`;

const CenterMeter = styled.div`
    position: absolute;
    top: 0;
    left: 50%;
    width: 0;
    height: 20px;
    border: 1px solid black;
`;

type ProgressBarProps = {
    pace: number
    needed: number
}

const ProgressBar = ({pace, needed}: ProgressBarProps) => {
    const binding = usePaceBar({pace, needed});
    return (
        <FullBar>
            <Progress color={binding.color} width={binding.width} />
            <CenterMeter />
        </FullBar>
    );
}

type UsePaceBarArgs = {
    pace: number
    needed: number
}

const usePaceBar = (args: UsePaceBarArgs) => {

    const percent = useMemo(() => {
        let pace = args.pace;
        if (args.pace > args.needed) {
            pace = pace - args.needed;
            pace = 2 * ((1 / (1 + Math.pow(Math.E, -pace / args.needed))) - 0.5); // sigmoid
            pace = pace * args.needed + args.needed;
        }
        return pace / args.needed * 50;
    }, [args.pace, args.needed]);

    const color = useMemo(() => percent >= 50 ? 'green' : '', [percent]);
    const width = useMemo(() => `${percent}%`, [percent]);
    return {
        color,
        width
    }
}

export default ProgressBar;