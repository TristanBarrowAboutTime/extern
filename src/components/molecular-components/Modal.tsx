import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';
import React from 'react';
import Button from '../atomic-components/Button';
import { ButtonType } from '../../types/ButtonType';
import Styles from '../../style/Styles';
import styled from 'styled-components';

type ModalProps = {
    title: string
    content: string
    buttons: {
        text: string,
        buttonType: ButtonType, 
        onClick:() => void,
    }[],
    closeModal: () => void
}

const Blackout = styled.div`
    position: fixed;
    display: flex;
    justify-content: center;
    align-items: flex-start;
    width: 100vw;
    height: 100vh;
    background-color: rgba(0,0,0,.1);
`;

const Container = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    width: 550px;
    margin-top: 75px;
    height: 200px;
    background-color: white;
`;

const Header = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-stretch;
    background-color: gray;
    border-top: 2px solid;
    height: 22px;
`;

const Title = styled.div`
    color: white;
`;

const Close = styled.div`
    margin-right: 5px;
    margin-top: 3;
    color: white;
    :hover {
        cursor: 'pointer';
    }
`;

const Body = styled.div`
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
`;

const Message = styled.div`
    margin-top: 24px;
`;

const ButtonSet = styled.div`
    margin-bottom: 16px;
`;

const Modal = ({
    title,
    content,
    buttons,
    closeModal,

}: ModalProps) => {
    return (
        <Blackout>
            <Container>
                <Header>
                    <div />
                    <Title>{title}</Title>
                    <Close>
                        <FontAwesomeIcon 
                            icon={faTimes} 
                            color={Styles.color.white} 
                            onClick={closeModal}
                        />
                    </Close>
                </Header>
                <Body>
                    <Message>{content}</Message>
                    <ButtonSet>
                        {buttons.map((button) => {
                            return (
                                <Button 
                                    key={button.text}
                                    buttonType={button.buttonType}
                                    onClick={button.onClick}
                                    text={button.text}
                                    styles={{marginLeft: 4, marginRight: 4}}
                                />
                            );
                        })}
                    </ButtonSet>
                </Body>
            </Container>
        </Blackout>
    );
}

export default Modal;