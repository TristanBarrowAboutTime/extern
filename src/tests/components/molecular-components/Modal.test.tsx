import React from 'react';
import renderer from 'react-test-renderer';
import Modal from '../../../components/molecular-components/Modal';
import { ButtonType } from '../../../components/atomic-components/Button';
import { faTimes } from '@fortawesome/free-solid-svg-icons';

describe('Modal', () => {

    it('renders correctly', () => {
        const modalMock = {
            title: 'title',
            content: 'content',
            buttons: [
                {
                    text: 'Button',
                    buttonType: ButtonType.NORMAL,
                    onClick: jest.fn()
                }
            ],
            closeModal: jest.fn()
        }

        const component = renderer.create(
            <Modal {...modalMock} />
        ).toJSON();

        expect(component).toMatchSnapshot();
    });

    it('gets its button clicked', () => {
        const buttonFunc = jest.fn();
        const closeModal = jest.fn();

        const modalMock = {
            title: 'title',
            content: 'content',
            buttons: [
                {
                    text: 'Button',
                    buttonType: ButtonType.NORMAL,
                    onClick: buttonFunc
                }
            ],
            closeModal: closeModal
        }

        const component = renderer.create(
            <Modal {...modalMock} />
        );

        component.root.findByProps({ key: 'Button' }).props.onClick();
        component.root.findByProps({ icon: faTimes }).props.onClick();

    });
})