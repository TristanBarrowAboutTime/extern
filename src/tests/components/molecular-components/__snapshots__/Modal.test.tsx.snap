// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<div
  className="sc-gtssRu eqKeKj"
>
  <div
    className="sc-dlnjPT fYNbyB"
  >
    <div
      className="sc-hKFyIo iyYxAx"
    >
      <div />
      <div
        className="sc-eCApGN dvxpGk"
      >
        title
      </div>
      <div
        className="sc-jSFkmK bIECDV"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          color="#FFFFFF"
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          onClick={[MockFunction]}
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div
      className="sc-gKAblj cSoOjD"
    >
      <div
        className="sc-iCoHVE bWLRIE"
      >
        content
      </div>
      <div
        className="sc-fujyUd hkeMBa"
      >
        <button
          className="sc-bdnylx cqoqID"
          disabled={false}
          onClick={[MockFunction]}
          style={
            Object {
              "marginLeft": 4,
              "marginRight": 4,
            }
          }
        >
          Button
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Not Implemented yet is true 1`] = `
<div
  className="sc-gtssRu eqKeKj"
>
  <div
    className="sc-dlnjPT fYNbyB"
  >
    <div
      className="sc-hKFyIo iyYxAx"
    >
      <div />
      <div
        className="sc-eCApGN dvxpGk"
      >
        title
      </div>
      <div
        className="sc-jSFkmK bIECDV"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          color="#FFFFFF"
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          onClick={[MockFunction]}
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div
      className="sc-gKAblj cSoOjD"
    >
      <div
        className="sc-iCoHVE bWLRIE"
      >
        content
      </div>
      <div
        className="sc-fujyUd hkeMBa"
      >
        <button
          className="sc-bdnylx cqoqID"
          disabled={false}
          onClick={[MockFunction]}
          style={
            Object {
              "marginLeft": 4,
              "marginRight": 4,
            }
          }
        >
          Button
        </button>
      </div>
    </div>
  </div>
</div>
`;
